cirq-core/cirq/transformers/routing/line_initial_mapper.py

```diff
@@ -0,0 +1,221 @@
+            if q0 not in self.partners:
+                self.partners[q0] = q1
+            if q1 not in self.partners:
+                self.partners[q1] = q0
+
+            if not (degree_lt_two(q0) and degree_lt_two(q1) and c0 != c1):
+                continue
+
+            # Make sure c0/q0 are for the largest component.
+            if len(circuit_graph[c0]) < len(circuit_graph[c1]):
+                c0, c1, q0, q1 = c1, c0, q1, q0
+
+            # copy smaller component into larger one.
+            c1_order = (
+                reversed(circuit_graph[c1])
+                if circuit_graph[c1][-1] == q1
+                else iter(circuit_graph[c1])
+            )
+            for q in c1_order:
+                if circuit_graph[c0][0] == q0:
+                    circuit_graph[c0].appendleft(q)
+                else:
+                    circuit_graph[c0].append(q)
+                component_id[q] = c0
+
+        return sorted([circuit_graph[c] for c in set(component_id.values())], key=len, reverse=True)
+
+    def initial_mapping(self, circuit: 'cirq.AbstractCircuit') -> Dict['cirq.Qid', 'cirq.Qid']:
+        """Maps disjoint lines of logical qubits onto lines of physical qubits.
+
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            a dictionary that maps logical qubits in the circuit (keys) to physical qubits on the
+            device (values).
+        """
+        qubit_map: Dict['cirq.Qid', 'cirq.Qid'] = {}
+        circuit_graph = self._make_circuit_graph(circuit)
+        self.center = nx.center(self.device_graph)[0]
+
+        def next_physical(current_physical: 'cirq.Qid') -> 'cirq.Qid':
+            # Greedily map to highest degree neighbor that that is available
+            sorted_neighbors = sorted(
+                self.device_graph.neighbors(current_physical),
+                key=lambda x: self.device_graph.degree(x),
+                reverse=True,
+            )
+            for neighbor in sorted_neighbors:
+                if neighbor not in self.mapped_physicals:
+                    return neighbor
+            # If cannot map onto one long line of physical qubits, then break down into multiple
+            # small lines by finding nearest available qubit to the physical center
+            return self._closest_unmapped_qubit(self.center)
+
+        pq = self.center
+        first_isolated_idx = len(circuit_graph)
+        for idx, logical_line in enumerate(circuit_graph):
+            if len(logical_line) == 1:
+                first_isolated_idx = idx
+                break
+
+            for lq in logical_line:
+                self.mapped_physicals.add(pq)
+                qubit_map[lq] = pq
+                # Edge case: if mapping n qubits on an n-qubit device should not call next_physical
+                # when finished mapping the last logical qubit else will raise an error
+                if len(circuit.all_qubits()) != len(self.mapped_physicals):
```

Comment:
`len(circuit.all_qubits()) ` is very slow! `circuit.all_qubits()` iterates over all operations in the circuit, and adds all qubits to a frozen set and returns that frozen set, without any caching. This is because `cirq.Circuit` is a mutable object and number of qubits in the circuit can change with time. 

Let's precompute the number of qubits at the beginning of the function once, using `sum(len(x) for x in circuit_graph)`, and re-use the value here.